### PR TITLE
Sync With Upstream

### LINKED ISSUE #41: Resync movement CLI with latest upstream aptos CLI
The `aptos`-turned-`movement` crate needs to be re-synced from the latest upstream to benefit from all the accumulated goodness and retain compatibility.

_Originally posted by @mzabaluev in https://github.com/movementlabsxyz/aptos-core/pull/30#discussion_r1669121333_

### PR DESCRIPTION
**Pull request by l-monninger**
Resolves #41

### CODE REVIEW

**Comment by mzabaluev** [comment]
(The diff view in GitHub is unusable, will post excerpts)

We used to need this impl for `FinalityView`: 

```diff
--- a/storage/storage-interface/src/lib.rs
+++ b/storage/storage-interface/src/lib.rs
@@ -479,58 +485,11 @@ pub trait DbReader: Send + Sync {
         state_key: &StateKey,
         version: Version,
     ) -> Result<(Option<StateValue>, SparseMerkleProof)> {
-        self.get_state_value_with_proof_by_version_ext(state_key, version)
+        self.get_state_value_with_proof_by_version_ext(state_key, version, 0)
             .map(|(value, proof_ext)| (value, proof_ext.into()))
     }
 }
 
-impl DbReader for Arc<dyn DbReader> {
-    fn get_read_delegatee(&self) -> &dyn DbReader {
-        &**self
-    }
-}
```

Otherwise, the changes look OK.

File: aptos-move/e2e-move-tests/Cargo.toml [hunk 1]
```diff
@@ -22,7 +22,7 @@ aptos-gas-profiling = { workspace = true }
 aptos-gas-schedule = { workspace = true, features = ["testing"] }
 aptos-language-e2e-tests = { workspace = true }
 aptos-package-builder = { workspace = true }
-aptos-types = { workspace = true }
```
Comment (mzabaluev) [inline aptos-move/e2e-move-tests/Cargo.toml:25]: Do we really need testing features in production code?

The "testing" feature requirement should be moved to the dev-dependencies section.

File: aptos-move/vm-genesis/Cargo.toml [hunk 2]
```diff
@@ -17,7 +17,7 @@ aptos-cached-packages =  { workspace = true }
 aptos-crypto = { workspace = true }
 aptos-framework =  { workspace = true }
 aptos-gas-schedule = { workspace = true } 
-aptos-types = { workspace = true }
```
Comment (mzabaluev) [inline aptos-move/vm-genesis/Cargo.toml:20]: Same as above.

File: config/src/config/identity_config.rs [hunk 3]
```diff
@@ -38,7 +38,15 @@ pub struct IdentityBlob {
...
-        Ok(serde_yaml::from_str(&fs::read_to_string(path)?)?)
+        let content = fs::read_to_string(path)?;
+        match serde_yaml::from_str(&content) {
+            Ok(identity_blob) => Ok(identity_blob),
+            Err(e) => {
+                eprintln!("Deserialization error: {:?}", e);
+                eprintln!("YAML content:\n{}", content);
+                Err(e.into())
+            },
+        }
```
Comment (mzabaluev) [inline config/src/config/identity_config.rs:49]: Looks like temporary debug code, could be replaced with upstream.

File: crates/indexer/src/models/move_tables.rs [hunk 4]
```diff
@@ -56,6 +57,10 @@ impl TableItem {
         transaction_version: i64,
         transaction_block_height: i64,
     ) -> (Self, CurrentTableItem) {
+        debug!(
+            "Decoding WriteTableItem: {:?}",
+            write_table_item
+        );
```
Comment (mzabaluev) [inline crates/indexer/src/models/move_tables.rs:63]: Debug code, could be removed from the delta.

File: execution/executor/src/block_executor.rs [hunk 5]
```diff
@@ -229,7 +229,7 @@ where
                         .start_timer();
                     info!("next_version: {}", parent_output.next_version());
                     CachedStateView::new(
-                        StateViewId::BlockExecution { block_id },
```
Comment (mzabaluev) [inline execution/executor/src/block_executor.rs:232]: This needs to be in as our workaround, right?

**Review by mzabaluev — APPROVED** [review]
The rebasing of our changes looks sensible.

Please switch the base to `monza` before merging.

File: aptos-move/e2e-move-tests/Cargo.toml [hunk 6]
```diff
@@ -22,7 +22,7 @@ aptos-gas-profiling = { workspace = true }
...
+aptos-types = { workspace = true, features = ["testing"] }
```
Comment (mzabaluev) [inline aptos-move/e2e-move-tests/Cargo.toml:25]: Do we really need testing features in production code?

The "testing" feature requirement should be moved to the dev-dependencies section.

File: aptos-move/vm-genesis/Cargo.toml [hunk 7]
```diff
@@ -17,7 +17,7 @@ aptos-cached-packages =  { workspace = true }
...
+aptos-types = { workspace = true, features = ["testing"] }
```
Comment (mzabaluev) [inline aptos-move/vm-genesis/Cargo.toml:20]: Same as above.

File: execution/executor/src/block_executor.rs [hunk 8]
```diff
@@ -229,7 +229,7 @@ where
...
+                        StateViewId::Miscellaneous,
```
Comment (mzabaluev) [inline execution/executor/src/block_executor.rs:232]: Does this need to remain as our workaround?

File: storage/aptosdb/src/db/include/aptosdb_writer.rs [hunk 9]
```diff
@@ -197,6 +199,76 @@ impl DbWriter for AptosDB {
             Ok(())
         })
     }
+
+    /// Revert a commit.
+    fn revert_commit(&self, ledger_info_with_sigs: &LedgerInfoWithSignatures) -> Result<()> {
```
Comment (mzabaluev) [inline storage/aptosdb/src/db/include/aptosdb_writer.rs:204]: This is cherry-picked from #28, not yet reviewed there. As I wrote this code, I'll leave it as an exercise for @0xmovses.

File: storage/aptosdb/src/db/include/aptosdb_writer.rs [hunk 10]
```diff
@@ -561,18 +688,6 @@ impl AptosDB {
...
-                .ledger_db
-                .metadata_db()
-                .get_latest_ledger_info_option()
-                .map_or(0, |li| li.ledger_info().next_block_epoch());
-            ensure!(
-                x.ledger_info().epoch() == current_epoch,
-                "Gap in epoch history. Trying to put in LedgerInfo in epoch: {}, current epoch: {}",
-                x.ledger_info().epoch(),
-                current_epoch,
-            );
```
Comment (mzabaluev) [inline storage/aptosdb/src/db/include/aptosdb_writer.rs:574]: I'm not sure this check really needs to be removed...

File: storage/aptosdb/src/db/include/aptosdb_writer.rs [hunk 11]
```diff
@@ -630,4 +745,42 @@ impl AptosDB {
 
         Ok(())
     }
+
+    // Update in-memory state of the database and the metrics before reverting.
+    // Note that any failures in persisting the revert should be treated as
+    // non-recoverable.
+    fn pre_revert(
```
Comment (mzabaluev) [inline storage/aptosdb/src/db/include/aptosdb_writer.rs:752]: This is also a cherry-pick from #28, not yet used anywhere, but will be once that PR lands.

File: storage/aptosdb/src/state_store/mod.rs [hunk 12]
```diff
@@ -549,15 +551,17 @@ impl StateStore {
     }
 
     pub fn reset(&self) {
-        let (buffered_state, smt_ancestors) = Self::create_buffered_state_from_latest_snapshot(
+        let lock = self.reset_lock();
+        lock.reset();
+    }
+
+    pub fn reset_lock(&self) -> ResetLock<'_> {
```
Comment (mzabaluev) [inline storage/aptosdb/src/state_store/mod.rs:558]: Another cherry-pick from #28; it looks like a needless refactoring now, but will come useful later.

File: testsuite/forge/Cargo.toml [hunk 13]
```diff
@@ -15,7 +15,7 @@ rust-version = { workspace = true }
 [dependencies]
 again = { workspace = true }
 anyhow = { workspace = true, features = ["backtrace"] }
-aptos = { workspace = true }
```
Comment (mzabaluev) [inline testsuite/forge/Cargo.toml:18]: This is a zombie file resurrected by merge. As the crate does not exist in the movement fork, could be removed.

File: testsuite/smoke-test/Cargo.toml [hunk 14]
```diff
@@ -14,7 +14,7 @@ rust-version = { workspace = true }
 
 [dependencies]
 anyhow = { workspace = true }
-aptos = { workspace = true, features = ["fuzzing"] }
+movement = { workspace = true, features = ["fuzzing"] }
```
Comment (mzabaluev) [inline testsuite/smoke-test/Cargo.toml:17]: This is a zombie file resurrected by merge. As the crate does not exist in the movement fork, could be removed.

File: testsuite/testcases/Cargo.toml [hunk 15]
```diff
@@ -14,7 +14,7 @@ rust-version = { workspace = true }
 
 [dependencies]
 anyhow = { workspace = true }
-aptos = { workspace = true, features = ["fuzzing"] }
+movement = { workspace = true, features = ["fuzzing"] }
```
Comment (mzabaluev) [inline testsuite/testcases/Cargo.toml:17]: This is a zombie file resurrected by merge. As the crate does not exist in the movement fork, could be removed.

**Comment by mzabaluev** [comment]
The changes look sensible now, apart from some minor issues. The only real concern that jumped at me is the added `todo!` macros, potentially introducing crashiness.

File: aptos-move/vm-genesis/Cargo.toml (hunk 7 above)
Comment (mzabaluev) [reply aptos-move/vm-genesis/Cargo.toml:20]: This is needed, it's too cumbersome to extract a subset used in vm-genesis, and it's some lazy-parsed JSON.
So let's keep this as a visible wart.

**Comment by mzabaluev** [comment]
The merge breaks the `movement` CLI (in `./crates/aptos`), because it updates many of its dependencies, and resurrects others in a broken form, by seemingly resolving "removed by us/changed by them" conflicts in favor of the usually incomplete set of files changed upstream. The conflicts in `Cargo.toml` are also resolved in favor of resurrecting these in

I think the sensible way forward is to completely check out the required crates from `main`, and incorporate #41 into this so that the `movement` CLI is up-to-date against its workspace dependencies. Unfortunately, this means partially overwriting some changes made for Monza. We can keep them in the `monza` branch and revisit if we decide to pick up work on that project again.

**Comment by l-monninger** [comment]
My mistake, I misused `--all-targets` I guess: https://github.com/rust-lang/cargo/issues/10958

File: crates/aptos/CHANGELOG.md [hunk 16]
```diff
@@ -1,92 +1,52 @@
-# Aptos CLI Changelog
+# Movement CLI Changelog
 
-All notable changes to the Aptos CLI will be captured in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and the format set out by [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+All notable changes to the Movement CLI will be captured in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and the format set out by [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
-## Unreleased
-- Add balance command to easily get account balances for APT currently
-
-## [3.4.1] - 2024/05/31
```
[the remark this answers is not in the document]
Comment (mzabaluev) [reply crates/aptos/CHANGELOG.md:8]: Re-syncing is done in #43.

**Comment by mzabaluev** [comment]
This should be good to merge now.

File: crates/aptos/CHANGELOG.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (andygolay) [reply crates/aptos/CHANGELOG.md:8]: Do you want to rename the associated `aptos` directory to `movement` ?

**Pull request merged**